### DISCUSSION TITLE
aeld with log argument - formerly aell

### DISCUSSION THREAD

**Discussion by Procyon-b** [uBlockOrigin/uAssets#17907] [Q&A] [answered]
I can't seem to make it work.
I'm even trying the examples provided in the wiki, but still nothing is displayed.
```
wikipedia.org##+js(aeld, { "type": "/.^/", "log": 2 })
```
> "Above filter will log all calls without defusing any of them (because type can't match _anything_)"

I've tested with FF 112.0.2 + uBO 1.49.0 and with chrome 109 + uBO 1.49.2
I also still have an old chrome 7x with uBO 1.46.0 where I've tested with `wikipedia.org##+js(aell)`

**aell** gives me a full listing of all eventlisteners (there are 79), while **aeld** shows nothing in FF and chrome.

I've checked in the logger that the filter is really applied.

What am I doing wrong?

**Comment by mapx-** [comment]
your filter is working fine on my end

**Reply by gorhill** [reply]
Yes, works fine as is here too.

@Procyon-b Are you filtering the output of the console?

**Reply by Procyon-b** [reply]
No filtering applied. I've tripled-checked.

Here is how it looks with aell
![aell](https://user-images.githubusercontent.com/40043075/235792479-58c0584e-4281-4d51-911b-15e3acde38a9.png)

And here is with aeld+log
![aeld-log](https://user-images.githubusercontent.com/40043075/235792493-c4aa4c7e-4ebe-438d-ae06-2a6310f98fb1.png)

**Reply by gorhill** [reply]
Sorry, I don't know what to say, it works here and I fail to see why it wouldn't work on your side:

![Screenshot from 2023-05-02 18-13-04](https://user-images.githubusercontent.com/585534/235797329-7dfc0ed0-27cd-4b1e-b512-0be60528088e.png)

**Reply by Procyon-b** [reply]
FYI I've created a new profile in FF, installed ubo, updated the lists, added `wikipedia.org##+js(aeld, { "type": "/.^/", "log": 2 })`, then loaded the wikipedia page. Still nothing displayed.

To make sure that something else (antivirus) isn't involved. Just for the fun, I've lauched a FF test profile that I haven't used for ages. It's running ubo 1.40.8.
**aell** displays the _ael_ calls.

edit:
I'm running windows 8.1  x64 on this computer.

**Reply by gorhill** [reply]
I can reproduce not-working with a new Firefox profile.

**Reply by gorhill** [reply]
Sorry, completely forgot: `log`/`debug` are not working on stable release of uBO, only on dev build. The idea is that I wanted to prevent giving tools to investigate running scriptlets to site authors.

**Reply by Procyon-b** [reply]
No problem. Now I have an explanation  :)

**Reply by krystian3w** [reply]
Only now it is difficult to get out of this situation as they find a wiki with information or understand how JS code works without reading discussions and wikis.

**Reply by gorhill** [reply]
Another reason is that I wanted to be sure that if stray filters with `log`/`debug` find their way into the lists, they won't cause issues with dev tools. This might have been the main reason.

Yes, website authors skilled with development tools will just install uBO dev build, and in the end nothing we can think of is going to stop them.

**Reply by Yuki2718** [reply]
Will log functionality be available for dev build only? I thought it was just at the moment and will come to stable too.

**Reply by gorhill** [reply, marked as answer]
Stays dev build only.

**Reply by Yuki2718** [reply]
> Yes, website authors skilled with development tools will just install uBO dev build, and in the end nothing we can think of is going to stop them.

How about also requirng `filterAuthorMode true`? Sure they can find, but better than nothing?

**Reply by gorhill** [reply]
What's a really convincing argument for having this available in stable build?

**Reply by Yuki2718** [reply]
No, I meant adding another barrier in dev build. But maybe site owner who found using dev build will too find the flag...

**Comment by gorhill** [comment]
Well it turns out a new way to pass extra parameters is working rather well I think, like with the new `sed` scriptlet, where pairs of extra parameters can be passed (position-independent), so I think this should be the way to go instead of the JSON approach. This would make using logging less cumbersome:

    wikipedia.org##+js(aeld, /.^/, , runAt, idle, log, 1)

**Reply by MasterKia** [reply]
`##+js(nosiif)` and `##+js(nostif)` activate logging, so maybe `##+js(aeld)` should also do that?

**Reply by gorhill** [reply]
`nostif` was designed before  the current way, I rather keep the current way and convert `nostif` to the current way, and have both `aeld` and `nostif` defuse all calls to `setTimeout` and `addEventListener` when no parameters provided.

**Reply by MasterKia** [reply]
> convert nostif to the current way

I don't know about other filter maintainers but I think `##+js(nostif)` to log and `##+js(nostif, /^/)` to defuse all calls are way more convenient to use and memorize than `##+js(nostif, /.^/, , log, 1)`.
I have to check the docs for `aeld` logging every time.

**Reply by gorhill** [reply]
> than ##+js(nostif, /.^/, , log, 1).

`##+js(nostif, , , log, 1)`

The new way allows different level of logging, and it allows to log only `click` types, etc.

**Reply by okiehsch** [reply]
>  are way more convenient to use and memorize

They are clearly easier to memorize:
I added
`! *##+js(aeld, { "type": "/.^/", "log": 2 })`
at the end of my filter list, that way I only have to remove the `!` if I want to do a basic log of any given site.

**Reply by peace2000** [reply]
What is the current correct way to log using `aeld`?

Is it:
`iltalehti.fi##+js(aeld, { "type": "/.^/", "log": 2 })`
or
`iltalehti.fi##+js(aeld, /.^/, , runAt, idle, log, 1)`

For some reason neither of these work for me on `https://www.iltalehti.fi/`

I'm running 1.49.3rc5

**Reply by peace2000** [reply]
I can't get those to work in any other websites either.

**Reply by gorhill** [reply]
iltalehti.fi##+js(aeld, , , log, 1)

When log is 1, logging occurs when both type and needle match, when log is 2, logging occurs when at least one matches, when log is 3, logging occurs for all calls to addEventListsener. Using `/.^/` with log 1 was preventing logging to occurs since type never matched.

---

To defuse nothing but log all call:

    iltalehti.fi##+js(aeld, /.^/, , log, 3)

But any type which very unlikely to match would also work:

    iltalehti.fi##+js(aeld, zzzz, , log, 3)

**Reply by gothic-bum** [reply]
> When log is 1, logging occurs when both type and needle match, when log is 2, logging occurs when at least one matches

Intuitively, shouldn't this be reversed so that `log, 1` is used when a single (i.e. one) type or needle matches, and `log, 2` be used when they both (i.e. two) match?

**Reply by gorhill** [reply]
If the goal is to make it easier to remember, we could just go with natural language:

    log, all
    log, match
    log, matchall

First log all the time, match or no match, the other only when there is at least one match, and the last when all conditions match.

I will leave it to filter maintainers to decide.

**Reply by mapx-** [reply]
@gorhill 

it seems 
`*##+js(aeld, , , log, 1)`
is breaking this site (games thumbnails are broken)
`https://steamcrackedgames.com/`

or the page is totally unresponsive (cannot edit "text string")
`https://regex101.com/r/VatKY3/1`

1.49.3rc5

**Reply by stephenhawk8054** [reply]
I personally prefer using words since it's indeed easier to remember for me.

**Reply by peace2000** [reply]
> gorhill
> 
> it seems `*##+js(aeld, , , log, 1)` is breaking this site (games thumbnails are broken) `https://steamcrackedgames.com/`
> 
> or the page is totally unresponsive `https://regex101.com/r/VatKY3/1`

I can't repro broken thumbnails or unresponsiveness with `*##+js(aeld, , , log, 1)`. Works fine on my end - with both FF and Chrome.

**Reply by mapx-** [reply]
on `regex101` I cannot edit the "text string" box content

**Reply by peace2000** [reply]
I also think that using real words is better.

This is very logical and easy to remember:
`log, all`

but how about changing the rest:

`log, match` -> `log, partialmatch`
`log, matchall` -> `log, fullmatch`

`match` and `matchall` still sound somewhat ambiguous to me. I can't fully figure out what they do unless I read the wiki.

**Reply by peace2000** [reply]
> on `regex101` I cannot edit the "text string" box content

Thanks for clarification. I can repro that issue as well.

Also, I can't click the three dot message menu open on GH. I couldn't quote your message without disabling that filter.

**Reply by mapx-** [reply]
`example.com##+js(aeld, , , log,)` => all 
`example.com##+js(aeld, , , log, 1)` => matching first parameter
`example.com##+js(aeld, , , log, 2)` => matching second parameter

**Reply by mapx-** [reply]
yeah, any single site is broken using `*##+js(aeld, , , log, 1)`
see also `https://megaup.net/1SXvU/Stars.Arisen.rar` => it starts loading but it remains in some waiting state

**Reply by gothic-bum** [reply]
Doesn't "type" and "pattern" default to *match-all*, thus making `*##+js(aeld, , , log, 1)` the same as `*##+js(aeld, /^/, /^/, log, 1)`? 

If so, wouldn't breakage with `*##+js(aeld, , , log, 1)` be expected since all types and patterns are being defused, which are then logged with `log, 1`?

**Reply by mapx-** [reply]
yea, it's a bit of confusion, I'd expect when log is present no defusing happens

**Reply by gorhill** [reply]
> yeah, any single site is broken using `*##+js(aeld, , , log, 1)`

This defuses everything, you need to add something to prevent matches:

    *##+js(aeld, _, , log, 3)

**Reply by peace2000** [reply]
>*##+js(aeld, _, , log, 3)

This filter blocks something on `iltalehti.fi`. It someway breaks GDPR related stuff. It removes a button from the bottom of the page that allows to change cookie settings, and prevents embedded from appearing.

Sample link:
https://www.iltalehti.fi/ulkomaat/a/9435695b-e455-4c18-8a2d-68e6e241c31f

With that filter:
https://github.com/uBlockOrigin/uAssets/assets/17256841/de021d61-ae6e-41df-bb1e-ce47ec0f0a4d
https://github.com/uBlockOrigin/uAssets/assets/17256841/856aa25c-a266-4af8-9792-380b8329dc13

Without that filter:
https://github.com/uBlockOrigin/uAssets/assets/17256841/3cc49451-a98b-46c1-84fd-2f5f956adb3c
https://github.com/uBlockOrigin/uAssets/assets/17256841/d1d761e4-8ecc-4be3-86f9-710e8da82491

What filter would really log everything without blocking anything?

**Reply by peace2000** [reply]
This is mentioned earlier `##+js(aeld, /.^/, , log, 3)` as logging everything but not defusing anything. It doesn't seem to break anything on `iltalehti.fi`.

The wiki however https://github.com/gorhill/uBlock/wiki/Resources-Library#addeventlistener-defuserjs- says:

```
In summary, if you need to only log all event listeners, use this filter:

example.com##+js(aeld, _, , log, 3)
```

But apparently this information is wrong?

**Reply by ghost** [reply]
Same issue as with replace node text? `_` matches in `alma_cmp_v2_consents_available_repetitive`? (plenty of events like this on this page)

**Reply by gorhill** [reply]
I can fix it the same I did for rmnt/rpnt

**Reply by gorhill** [reply]
https://github.com/gorhill/uBlock/commit/d32204f9842091261afd5bf3eea90573f2058ece

**Reply by ghajini** [reply]
`example.com##+js(aeld, /.^/, , log, 2)` will log all calls without defusing?

**Reply by gorhill** [reply]
Nothing changed when regexes. Change is about reducing the likelihood of `example.com##+js(aeld, _, , log, 3)` from defusing by requiring an exact match for non-regex patterns -- that was the original intent of the scriptlet, probably lost in some refactoring.

**Reply by peace2000** [reply]
>https://github.com/gorhill/uBlock/commit/d32204f9842091261afd5bf3eea90573f2058ece

Thanks, the issue was fixed.

**Reply by peace2000** [reply]
@gorhill 

How can I nowadays log every event on a page by using aeld?

The current, documented filter:

`example.com##+js(aeld, _, , log, 3)`

Doesn't log anything anymore.

**Reply by garry-ut99** [reply]
None of the others from Wiki work either anymore:
`wikipedia.org##+js(aeld, /mouse/, /.^/, log, 2)`
`wikipedia.org##+js(aeld, , /.^/, log, 2)`

Only such one from Wiki works:
`wikipedia.org##+js(aeld, , , log, 1)` but it will not only log but also defuse

Seems "logging only" feature is broken completely.

**Reply by stephenhawk8054** [reply]
@peace2000 Does this work?

```adb
example.com##+js(aeld)
```

**Reply by garry-ut99** [reply]
Hmm, not differently than the one which I already provided, both of them are equal and only log all calls, not defuse, as both of them only says "called" in Logger:
`wikipedia.org##+js(aeld, , , log, 1)` 
`example.com##+js(aeld)`

despite [Wiki](https://github.com/uBlockOrigin/uBlock-issues/wiki/Resources-Library#addeventlistener-defuserjs-) says  about the one I provided: "The forth filter will log and defuse _all_ calls to `addEventListener()`."

However, now, I'm not sure to which filter that statement really is reffering, as mathematically the forth is: `jpvhub.com##+js(aeld, click, popMagic, runAt, idle)`, which doesn't seem to be the one, the statement is reffering to, as it doesn't "defuse and log _all_ calls", hence I assumed "forth" was a typo and it should be "third".

But on the other hand the third one (the one I provided) doesn't defuse anything, only logs, even `wikipedia.org##+js(aeld, , , )` doesn't defuse anything, only logs, so I don't know why the statement says it defuses, perhaps another typo ? But two typos in one statement ? 

Looking at Wiki commits I see now "third" changed to "forth" by this [commit](https://github.com/uBlockOrigin/uBlock-issues/wiki/Resources-Library/_compare/e7c90b103def88bb190700957d887e361e7b52af...e79da8583ae18439a5b11a95c82fc2379b0f7c93#diff-46c88adfa593709bcf62f320faae0cdaccf08546c4536e0b493cdc8130e791b2R268) by @MasterKia, now I wonder which commit was a typo: adding "third", or changing  "third" to "forth", or maybe the statement is wrong, as it doesn't correctly describe neither 'third" nor "forth" filter, it seems someone needs to clean up Wiki about AddEvent Listener-defuser....or maybe there is a bug in the scriptlet as "first" and "second" don't log anything anymore,  despite their descriptions in Wiki say they should, or maybe descriptions are obsolete, or maybe this... ,or maybe that..., wow, enough.

**Reply by peace2000** [reply]
All of these seem to work for me, I see events in the logger :

```
example.com##+js(aeld)
example.com##+js(aeld, , , log, 1)
example.com##+js(aeld, , , log, 2)
example.com##+js(aeld, , , log, 3)
```

And none of them has "prevented" text so I assume nothing is being defused.

Probably wiki should be updated but I'm not comfortable to do so because I'm not fully sure how to scriptlet is intended to work

**Reply by peace2000** [reply]
Okay this filter:

`youtube.com##+js(aeld)`

Causes Youtube not to load at all. Logger doesn't say it has prevented something.

**Reply by Yuki2718** [reply]
Yeah, reproduced.

**Reply by gorhill** [reply]
The code at <https://github.com/gorhill/uBlock/blob/d8a77f38ceac203c8f146213e40b976e55ccfbe3/assets/resources/scriptlets.js#L1696-L1701> does show that when no parameters are present, there is logging and the call is not prevented.

**Reply by ghajini** [reply]
worth of creating proper issue here https://github.com/uBlockOrigin/uBlock-issues/issues
- i noticed when i go here eg `https://www.youtube.com/shorts/GxUDZK9dQaA` , i see broken ui with `youtube.com##+js(aeld)`
![image](https://github.com/uBlockOrigin/uAssets/assets/20338483/1f9d360e-93c2-49d7-8735-3b180c3d131a)

- without filter
![image](https://github.com/uBlockOrigin/uAssets/assets/20338483/9c03d4d5-16d4-4a4d-a22f-811d76f56091)

**Reply by gorhill** [reply]
Not uBO specific, same issue with AdGuard.

**Reply by gorhill** [reply]
If I remove the logging of the element id on which the listener was added, the issue goes away. It seems Youtube may be trapping the access to the `id` property of an element and this causes failure somewhere.

**Reply by gorhill** [reply]
So the logging of the element details was added as per request: <https://github.com/uBlockOrigin/uBlock-issues/issues/3061#issuecomment-1899042062>.

Accessing the `id` is causing an issue on Youtube, presumably because somewhere in Youtube's huge pile of JS, there is a trap which act on accessing the `id` of an element. I tried various things and somehow replacing:

        if (  elem.id !== '' ) { parts.push(`#${CSS.escape(elem.id)}`); }

With:

        const id = String(elem.id);
        if ( id !== '' ) { parts.push(`#${CSS.escape(id)}`); }

Ends up solving the issue -- kind of a voodoo-esque solution...

---

https://github.com/gorhill/uBlock/commit/8eb3b19c69e5960da2868095d82dc9403302c478

**Reply by garry-ut99** [reply]
> peace2000 : https://github.com/uBlockOrigin/uAssets/discussions/17907#discussioncomment-9848290 @ gorhill How can I nowadays log every event on a page by using aeld? The current, documented filter:  `example.com##+js(aeld, _, , log, 3)` Doesn't log anything anymore.

I've discovered a more up-to-date  (02.2024) `aeld` log related subdiscussion in another, semi-related discussion : https://github.com/uBlockOrigin/uBlock-issues/discussions/3096#discussioncomment-8338944 which says that `_` and `log` have become obsolete ways of logging scriptlets and that `_` doesn't work anymore (though `log` still does sometimes) and that the current way is `##+(aeld)` which prints all logs, but doesn't defuse anything (will print "called" only). 

> peace2000 : Probably wiki should be updated but I'm not comfortable to do so because I'm not fully sure how to scriptlet is intended to work

Yes, "first" and "second" `log` examples from WiKi don't work anymore due to being deprecated, and the "third" one which still works, has been made redundant by the current one (`##+js(aeld)`), and seems to work only by accident: as empty/incorrect/surplus parameters are being ignored/skipped, what results in behavior equivalent to `##+js(aeld)`, in all following examples (still more examples can be crafted):
```
wikipedia.org##+js(aeld, , , )
wikipedia.org##+js(aeld, , , , , )
wikipedia.org##+js(aeld, , , x, y, z)
wikipedia.org##+js(aeld, , , x, y, z, log, 3)
wikipedia.org##+js(aeld, , , a, b, c, log, 9)
wikipedia.org##+js(aeld, _, , log, 3)
wikipedia.org##+js(aeld, , , log, 3)
```

**Comment by MasterKia** [comment]
> so I think this should be the way to go instead of the JSON approach

Will the JSON approach be deprecated?

**Reply by gorhill** [reply]
I don't know. If eventually nobody uses it, probably.

**Reply by MasterKia** [reply]
https://github.com/gorhill/uBlock/commit/1d9ee00498584b8370840de297e748295b6f472f

I think `set-constant` and `acis` don't support the new way of passing optional parameters (They don't have `get-extra-args.fn` as dependency).

`joysound.com##+js(set, document.body.oncopy, null, 3)`
=>
`joysound.com##+js(set, document.body.oncopy, null, runAt, 3)`

**Reply by gorhill** [reply]
I will work on this in next dev cycle.
